deep_qa/training/models.py

```diff
@@ -146,6 +152,217 @@ def _make_predict_function(self):
+            names of the outputs of ``f`` and the list of display names of the outputs of ``f_val``.
+        initial_epoch: int, optional (default = 0).
+            The epoch at which to start training (useful for resuming a previous training run).
+
+        Returns
+        -------
+        A Keras ``History`` object.
+
+        """
+        do_validation = False
+        if val_f and val_ins:
+            do_validation = True
+            if verbose:
+                print('Train on %d samples, validate on %d samples' %
+                      (ins[0].shape[0], val_ins[0].shape[0]))
+
+        if ins and hasattr(ins[0], 'shape'):
+            num_train_samples = ins[0].shape[0]
+        else:
+            # May happen if we are running `fit` without Numpy input data,
+            # i.e. if all inputs to the models are data tensors
+            # instead of placeholders.
+            # In that case we will run `fit` over a single batch.
+            num_train_samples = batch_size
+            verbose = 2
+        index_array = numpy.arange(num_train_samples)
+        out_labels = out_labels or []
+        callbacks, callback_model = self._prepare_callbacks(callbacks, val_ins, epochs, batch_size,
+                                                            num_train_samples, callback_metrics,
+                                                            do_validation, verbose)
+
+        for epoch in range(initial_epoch, epochs):
+            callbacks.on_epoch_begin(epoch)
+            if shuffle == 'batch':
+                index_array = _batch_shuffle(index_array, batch_size)
+            elif shuffle:
+                numpy.random.shuffle(index_array)
+
+            batches = _make_batches(num_train_samples, batch_size)
+            epoch_logs = {}
+            for batch_index, (batch_start, batch_end) in enumerate(batches):
+                batch_ids = index_array[batch_start:batch_end]
+                try:
+                    if isinstance(ins[-1], float):
+                        # Do not slice the training phase flag.
+                        ins_batch = _slice_arrays(ins[:-1], batch_ids) + [ins[-1]]
+                    else:
+                        ins_batch = _slice_arrays(ins, batch_ids)
+                except TypeError:
+                    raise TypeError('TypeError while preparing batch. '
+                                    'If using HDF5 input data, '
+                                    'pass shuffle="batch".')
+
+                # Here is the main difference between a single gpu model and one split
+                # across multiple gpus. In our multiple gpu model, all of the inputs
+                # are replicated num_gpus times, so we need to split our large batch
+                # into the corresponding sets of smaller batches for each model.
+                if self.num_gpus > 1:
+
+                    def multi_gpu_batch(variable_list):
```

Comment:
You could pull this function definition out of the loop since it doesn't use any closures.  I'm not sure, but maybe there is some overhead to including the function definition in a loop?